memcached-operator/pkg/controller/memcached/memcached_controller.go

```diff
@@ -208,44 +196,70 @@ func (r *ReconcileMemcached) deploymentForMemcached(m *cachev1alpha1.Memcached)
 				},
 				Spec: corev1.PodSpec{
 					Containers: []corev1.Container{{
-						Image:   "memcached:1.4.36-alpine",
-						Name:    "memcached",
-						Command: []string{"memcached", "-m=64", "-o", "modern", "-v"},
-						Ports: []corev1.ContainerPort{{
-							ContainerPort: 11211,
-							Name:          "memcached",
-						}},
+						Name:    name,
+						Image:   image,
+						Ports:   ports,
+						Command: command,
 					}},
 				},
 			},
-		},
+		}
+
+		return
+	}
+
+	if *deployment.Spec.Replicas != replicas {
+		reqLogger.Info("Fixing replica count in Deployment", "From", deployment.Spec.Replicas, "To", replicas)
+		deployment.Spec.Replicas = &replicas
+	}
+	if !reflect.DeepEqual(deployment.Spec.Template.ObjectMeta.Labels, ls) {
+		reqLogger.Info("Fixing labels in Deployment Pod template", "From", deployment.Spec.Template.ObjectMeta.Labels, "To", ls)
+		deployment.Spec.Template.ObjectMeta.Labels = ls
+	}
+	if deployment.Spec.Template.Spec.Containers[0].Name != name {
+		reqLogger.Info("Fixing container name in Deployment Pod template", "From", deployment.Spec.Template.Spec.Containers[0].Name, "To", name)
+		deployment.Spec.Template.Spec.Containers[0].Name = name
+	}
+	if deployment.Spec.Template.Spec.Containers[0].Image != image {
+		reqLogger.Info("Fixing container image in Deployment Pod template", "From", deployment.Spec.Template.Spec.Containers[0].Image, "To", image)
+		deployment.Spec.Template.Spec.Containers[0].Image = image
+	}
+	if !reflect.DeepEqual(deployment.Spec.Template.Spec.Containers[0].Ports, ports) {
+		reqLogger.Info("Fixing container ports in Deployment Pod template", "From", deployment.Spec.Template.Spec.Containers[0].Ports, "To", ports)
+		deployment.Spec.Template.Spec.Containers[0].Ports = ports
 	}
```

Comment:
In many scenarios, I mean for many business logics would be great to ensure all and/or many specs. However, is a good idea we lead the idea that they should be checked always by implementing all possibilities here? Should really be a good practice have all these conditions in the updateOrCreate func? Is not the complexity of this func too high? 
Would not better has a func to check each spec instead of?

Reply:
I agree the code sprawl is not worth it just for log messages that will rarely be looked at. It's better to just generate the desired Spec and overwrite the existing one if it's not equal. If there's really a need to log what changed, perhaps make a dedicated comparison function that'll take the old and new Spec and log what's not the same. That would make it clear that it's debugging info for humans, and not code that is required to make a successful and correct operator.

Reply:
@grdryn I think the simple overwriting of individual fields that you have in your last alternative is a good way to go.

In the steady-state case, nothing would change, so theoretically no API call would be made to the server (that's one of the key aspects of the example, IIRC). I'm pretty sure that overwriting the spec would result in an extraneous API call, and we want to avoid that regardless of whether the resourceVersion changes.

I actually kind of like the `CreateOrUpdate` approach being used in the example because it makes it immediately clear that one of the main properties of the `Reconcile()` method is that it is idempotent.

Reply:
Hi @joelanford, 

My problem with is not exactly the usage of the func  `CreateOrUpdate` but how it has been used. For example,  
 at in this point, after the @grdryn improvements which make it very better, the complexity of this func still too high (+ the 57 lines).  So, IMHO each check should be encapsulated for example.